modules/gapi/src/backends/plaidml/gplaidmlbackend.cpp

```diff
@@ -0,0 +1,286 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+
+#ifdef HAVE_PLAIDML
+
+#include "precomp.hpp"
+
+#include <ade/util/algorithm.hpp>
+#include <ade/util/range.hpp>
+#include <ade/util/zip_range.hpp>
+#include <ade/typed_graph.hpp>
+
+#include <opencv2/gapi/gcommon.hpp>
+#include <opencv2/gapi/util/any.hpp>
+#include <opencv2/gapi/gtype_traits.hpp>
+
+#include "compiler/gobjref.hpp"
+#include "compiler/gmodel.hpp"
+
+#include "backends/plaidml/gplaidmlbackend.hpp"
+#include "backends/plaidml/plaidml_util.hpp"
+
+#include "api/gbackend_priv.hpp" // FIXME: Make it part of Backend SDK!
+
+using GPlaidMLModel = ade::TypedGraph
+    < cv::gimpl::PlaidMLUnit
+    , cv::gimpl::Protocol
+    >;
+
+// FIXME: Same issue with Typed and ConstTyped
+using GConstGPlaidMLModel = ade::ConstTypedGraph
+    < cv::gimpl::PlaidMLUnit
+    , cv::gimpl::Protocol
+    >;
+
+namespace
+{
+    class GPlaidMLBackendImpl final: public cv::gapi::GBackend::Priv
+    {
+        virtual void unpackKernel(ade::Graph            &graph,
+                                  const ade::NodeHandle &op_node,
+                                  const cv::GKernelImpl &impl) override
+        {
+            GPlaidMLModel gm(graph);
+            auto plaidml_impl = cv::util::any_cast<cv::GPlaidMLKernel>(impl.opaque);
+            gm.metadata(op_node).set(cv::gimpl::PlaidMLUnit{plaidml_impl});
+        }
+
+        virtual EPtr compile(const ade::Graph &graph,
+                             const cv::GCompileArgs &,
+                             const std::vector<ade::NodeHandle> &nodes) const override
+        {
+            return EPtr{new cv::gimpl::GPlaidMLExecutable(graph, nodes)};
+        }
+   };
+}
+
+cv::gapi::GBackend cv::gapi::plaidml::backend()
+{
+    static cv::gapi::GBackend this_backend(std::make_shared<GPlaidMLBackendImpl>());
+    return this_backend;
+}
+
+void cv::gimpl::GPlaidMLExecutable::initBuffers(const std::vector<cv::gimpl::Data>& data,
+                                                std::vector<plaidml::exec::Binding>& bindings)
+{
+
+    // NB: This is necessary because we keep a pointer to bindings elements to buffer_map
+    // In order to them to remain valid it's required to prevant reallocation
+    bindings.reserve(data.size());
+    for (const auto& d : data)
+    {
+        GAPI_Assert(d.meta.index() == cv::GMetaArg::index_of<cv::GMatDesc>() &&
```

Comment:
`holds_alternative` is the right way to check this. Comparing index (which is just plain `int`) is not that reliable. What if `d.meta` is not `GMetaArg` anymore?

Reply:
Fixed